### DISCUSSION TITLE
Thermostat configuration

### DISCUSSION THREAD

**Discussion by fabnavigator** [zwave-js/node-red-contrib-zwave-js#230] [Q&A]
Unless I set the values from NR, I don't see the current configuration values for my thermostats.

In the screenshot below there only are values for the first three configuration parameters, those that I set with the interface. I was expecting to see all of the current configuration values.

https://devices.zwave-js.io/?jumpTo=0x0039:0x0011:0x0008:0.0

![Untitled](https://user-images.githubusercontent.com/16625904/178295493-578fc13a-afca-44b6-bf29-4242d7b9c7e0.png)

**Comment by marcus-j-davies** [comment]
If you double click the title of a config param - and check if it has a value (please send screenshot)

if it has - its a strange issue on the UI I need to address 😅 
if it hasn't  - its likely the interview, wasn't 100% successful ( clicking the handshake button for that node will reinterview it) - refresh after the interview is complete. 

![image](https://user-images.githubusercontent.com/55892693/178303639-f44db63f-000d-42e2-bb0d-f610257b8cbb.png)

**Reply by fabnavigator** [reply]
@marcus-j-davies 

The good news is that the values were blank. The bad news is that after I re-interviewed the node, all of the values are now blank.

**Reply by fabnavigator** [reply]
This is only an issue with the configuration values. Everything else (i.e., thermostat mode, setpoint, etc.) is showing values.

**Reply by marcus-j-davies** [reply]
Strange!

That's the whole point of an interview.
could you re-interview the device again, but enable logging before you do, and attach the log file here?

the same interview protocol is used for all devices, and is to spec. so not sure why you are not getting values - the log file will tell us.

**Reply by fabnavigator** [reply]
![Untitled](https://user-images.githubusercontent.com/16625904/178364352-00c922f6-50c6-4c2c-a8aa-87da43d4c558.png)

**Reply by fabnavigator** [reply]
[log.txt](https://github.com/zwave-js/node-red-contrib-zwave-js/files/9087812/log.txt)

**Reply by fabnavigator** [reply]
In the case below shouldn't the normalizedObject.label be "Heating"? As it is, I can't use the normalized object becasue I don't know which setpoint it's for.

![Untitled](https://user-images.githubusercontent.com/16625904/178372749-59f0159d-75e7-4a0c-99d1-fb6ea4291213.png)

**Comment by marcus-j-davies** [comment]
@fabnavigator 

 - **Your configuration interview log.**
Shows some timeouts and / or  S2 encapsulation errors when querying the config values.
Is this a buggy device when using S2 or is it a driver problem 🤷 

   Could I ask you raise a bug on the drivers repo, and attach this same log file, you attached here, just make reference to the config values are not being fetched during interview (along with this log file) - be sure to use the bug template, Unless @AlCalzone can answer it here. But I know he is on a well deserved break currently.
   
   
  - **In the case below shouldn't the normalizedObject.label be "Heating"? As it is, I can't use the normalized object becasue I don't know which setpoint it's for.**
  
    Yes, it should.
    Could you please run this, and send me the response?
    or paste the object on this [issue (232)](https://github.com/zwave-js/node-red-contrib-zwave-js/issues/232)
    
```javascript
    let Message = {
    "payload": {
        "mode": "ValueAPI",
        "node": 67,
        "method": "getValueMetadata",
        "params": [{commandClass:67, endpoint:0, property:"setpoint", propertyKey:1}]
    }
}
return Message
  
```

**Reply by fabnavigator** [reply]
Before I get too far into this, I'm going to try a couple of different thermostats. I will have a couple of TRANE 14942771 thermostats in a couple of days, and I'm still within the return window for the Honeywell TH6320ZW2003 thermostats. Many more, and much better reviews on the TRANE thermostats. Not to mention, the TRANE thermostats are close to 1/3 of the cost.

**Reply by fabnavigator** [reply]
I'm waiting for the new thermostats, but in the meantime I installed two new dimmer switches. When I used S2 security I ran into the same situation with the interview, and there were no configuration values. I excluded them and then included them unsecured, and everything worked right. So, I excluded and included the thermostats w/o security, and they interview correctly also. I can see all of the settings when I look at the configuration. Seems like something is going wrong when S2 security is being used.

**Reply by marcus-j-davies** [reply]
Very Strange.

I have not been able to reproduce this (neither has it been reported by other implementers of the same driver)
 - Home Assistant
 - io.broker
 
 I am currently using zwave-js **9.4.0**
 
 It's easy for me to say you just have buggy devices, but that is the wrong approach.
 my suggestion is, to capture an interview of these devices, into 4 separate log files.
 
  - device 1 (S2 interview and non S2)
  - device 2 (S2 interview and non S2)
  
and log it on the divers repo using the bug template.

1 difference however, is that you the TCP connection, but I really cant see why that will be the reason (and only on the Configuration CC) 🤷

**Reply by fabnavigator** [reply]
Just so I get this in the right place, please send me a link to the drivers repo.

**Reply by marcus-j-davies** [reply]
https://github.com/zwave-js/node-zwave-js/issues/new?assignees=&labels=&template=bug_report.yml

**Reply by AlCalzone** [reply]
Something is weird here with how the device responds to some requests, especially when S2 + transport encapsulation is used together. Roughly this happens:
```
-> Driver queries param information (with S2)
<- Device responds (with S2, split over 2 messages)
-> Driver confirms that both messages were received
<- Device begins another session (no idea with what)
-> at the same time, Driver queries the next param
```

The last two actions cause a collision in the S2 state which doesn't get resolved immediately since both parties retry their own command, which collides again.

I think this can be avoided for this device by disabling the info and name queries using these compat flags:
https://zwave-js.github.io/node-zwave-js/#/config-files/file-format?id=skipconfigurationnamequery
https://zwave-js.github.io/node-zwave-js/#/config-files/file-format?id=skipconfigurationinfoquery

**Reply by fabnavigator** [reply]
@AlCalzone Thank you for looking into this. I'm not sure how to set those flags using @marcus-j-davies software.

**Reply by marcus-j-davies** [reply]
Hi @fabnavigator,

Can you send me the DB link, and I'll add the flags shortly, and send you instructions.
it will make use of the **Custom CFG Dir** setting on the controller node.

**Reply by fabnavigator** [reply]
https://devices.zwave-js.io/?jumpTo=0x0039:0x0011:0x0008:0.0

**Reply by fabnavigator** [reply]
@marcus-j-davies Any status on when you will send me these instructions?

**Reply by marcus-j-davies** [reply]
Give me 30 Minutes

**Reply by marcus-j-davies** [reply]
@fabnavigator 

 - Create a directory on your system: lets say **/home/pi/zwave-custom-configs**
 - Place the contents of this zip file in there
 - On the controller node set **Custom CFG Dir** to the folder we created -> **/home/pi/zwave-custom-configs**
 - Deploy the changes
 - Reinterview this node
 
[th6320zw.zip](https://github.com/zwave-js/node-red-contrib-zwave-js/files/9151635/th6320zw.zip)

**Reply by fabnavigator** [reply]
@marcus-j-davies Thank you.
It's not finding the folder. I'm using Docker, you know. I created the config folder in my node-red directory, and didn't provide a path in the configuration node. I thought that would work because the zwave-js.log gets created in the node-red directory so I assume that the root from within the docker container is node-red.

**Reply by fabnavigator** [reply]
On a slight tangent, I excluded and then included the thermostat with S2 security and got a thumbs up in the GUI, indicating the interview was successful. But the log looks like this:
```
2022-07-20T16:08:48.646Z CNTRLR   [Node 107] Timed out while waiting for a response from the node (ZW0201)
2022-07-20T16:08:50.759Z CNTRLR   [Node 107] Timed out while waiting for a response from the node (ZW0201)
2022-07-20T16:08:51.831Z CNTRLR   [Node 107] Timed out while waiting for a response from the node (ZW0201)
2022-07-20T16:08:51.831Z CNTRLR   [Node 107] Querying parameter #2 information timed out, skipping interview...
2022-07-20T16:08:52.906Z CNTRLR   [Node 107] Timed out while waiting for a response from the node (ZW0201)
2022-07-20T16:08:52.907Z CNTRLR   [Node 107] Querying supported notification types timed out, skipping interview
                                  ...
2022-07-20T16:08:53.979Z CNTRLR   [Node 107] Timed out while waiting for a response from the node (ZW0201)
2022-07-20T16:08:55.050Z CNTRLR   [Node 107] Timed out while waiting for a response from the node (ZW0201)
2022-07-20T16:08:55.076Z CNTRLR   Failed to execute controller command after 1/3 attempts. Scheduling next try i
                                  n 100 ms.
2022-07-20T16:08:55.235Z CNTRLR   Failed to execute controller command after 1/3 attempts. Scheduling next try i
                                  n 100 ms.
2022-07-20T16:08:56.448Z CNTRLR   [Node 107] Timed out while waiting for a response from the node (ZW0201)
```
You can see that it skipped the interview.

**Reply by marcus-j-davies** [reply]
> It's not finding the folder. I'm using Docker, you know. I created the config folder in my node-red directory, and didn't provide a path in the configuration node. I thought that would work because the zwave-js.log gets created in the node-red directory so I assume that the root from within the docker container is node-red.

it needs an absolute path to a folder (that the driver can read), how that is done in docker. - I have no idea.

> On a slight tangent, I excluded and then included the thermostat with S2 security and got a thumbs up in the GUI, indicating the interview was successful. But the log looks like this:

A request for information on a CC or config param that does not respond is considered unsupported, failing the interview will happen if catastrophic communication occurs, a request for a single piece of information is not a failure - its just considered unsupported.  

In NR - I fail the interview, if the event of **interview failed** of the driver occurs on the `isFinal` attempt
https://github.com/zwave-js/node-red-contrib-zwave-js/blob/b21cf7a4bb65d78a9ad8a8d25cf0e2607c4f0e2d/zwave-js/zwave-js.js#L2050

The thumb will be displayed after the node's `ready` event has fired - which usually occurs after a completed interview, so will not be displayed if the interview failed (multiple  catastrophic communication occurs)
 
If I have that right @AlCalzone ?

**Reply by fabnavigator** [reply]
I added "/data" before the folder name and that worked. I was able to include both thermostats with S2 security, and I can see all of the configuration values. Big improvement. Thank you. The only issue I see is that after the interview, the temperature scale reverts to "Celsius".

Is this fix something that could be applied to other devices, or is it specific to a thermostat?

**Reply by marcus-j-davies** [reply]
I currently don't expose the options to change it (but they do exist in the driver).
will add them in V8 (7.2.0 has morphed into V8)

https://github.com/zwave-js/node-red-contrib-zwave-js/issues/242

**Reply by fabnavigator** [reply]
Sounds like a plan. When that is available do I have to change something to get that?

**Reply by marcus-j-davies** [reply]
It will just be a controller option like setting the debug level as an example

**Reply by fabnavigator** [reply]
Nice.

**Reply by AlCalzone** [reply]
If the flags worked, we should probably upstream the change.

**Reply by marcus-j-davies** [reply]
Yup,

Already have it in waiting, Will PR soon.
